LLama/Native/SafeLLamaSamplerHandle.cs

```diff
@@ -0,0 +1,593 @@
+        [DllImport(NativeApi.libraryName, CallingConvention = CallingConvention.Cdecl)]
+        static extern uint llama_sampler_get_seed(IntPtr smpl);
+    }
+
+    #region add/remove samplers
+    /// <summary>
+    /// Create a new sampler chain
+    /// </summary>
+    /// <param name="params"></param>
+    /// <returns></returns>
+    public static SafeLLamaSamplerChainHandle Create(LLamaSamplerChainParams @params)
+    {
+        return llama_sampler_chain_init(@params);
+
+        [DllImport(NativeApi.libraryName, CallingConvention = CallingConvention.Cdecl)]
+        static extern SafeLLamaSamplerChainHandle llama_sampler_chain_init(LLamaSamplerChainParams @params);
+    }
+
+
+    /// <summary>
+    /// Clone a sampler stage from another chain and add it to this chain
+    /// </summary>
+    /// <param name="src">The chain to clone a stage from</param>
+    /// <param name="index">The index of the stage to clone</param>
+    public void AddClone(SafeLLamaSamplerChainHandle src, int index)
+    {
+        if (index < 0 || index >= Count)
+            throw new ArgumentOutOfRangeException(nameof(index));
+
+        llama_sampler_chain_add(
+            this,
+            llama_sampler_clone(
+                llama_sampler_chain_get(src, index)
+            )
+        );
+
+        [DllImport(NativeApi.libraryName, CallingConvention = CallingConvention.Cdecl)]
+        static extern IntPtr llama_sampler_clone(IntPtr chain);
+    }
+
+    /// <summary>
+    /// Remove a sampler stage from this chain
+    /// </summary>
+    /// <param name="index"></param>
+    /// <exception cref="ArgumentOutOfRangeException"></exception>
+    public void Remove(int index)
+    {
+        if (index < 0 || index >= Count)
+            throw new ArgumentOutOfRangeException(nameof(index));
+
+        llama_sampler_free(llama_sampler_chain_remove(this, index));
+
+        [DllImport(NativeApi.libraryName, CallingConvention = CallingConvention.Cdecl)]
+        static extern IntPtr llama_sampler_chain_remove(SafeLLamaSamplerChainHandle chain, int i);
+    }
+
+
+    /// <summary>
+    /// Add a sampler which picks the most likely token.
+    /// </summary>
+    /// <returns></returns>
+    public void AddGreedySampler()
+    {
+        llama_sampler_chain_add(this, llama_sampler_init_greedy());
+
+        [DllImport(NativeApi.libraryName, CallingConvention = CallingConvention.Cdecl)]
+        static extern IntPtr llama_sampler_init_greedy();
+    }
+
+    /// <summary>
+    /// Add a sampler which picks from the probability distribution of all tokens
+    /// </summary>
+    /// <returns></returns>
+    public void AddDistributionSampler(uint seed)
+    {
+        llama_sampler_chain_add(this, llama_sampler_init_dist(seed));
+
+        [DllImport(NativeApi.libraryName, CallingConvention = CallingConvention.Cdecl)]
+        static extern IntPtr llama_sampler_init_dist(uint seed);
+    }
```

Comment:
Individual samplers which need a seed now accept a seed parameter just for that sampler. For example `llama_sampler_init_dist(uint seed);`.